### PR TITLE
Fix minimum requirement logic in teacher application

### PR DESCRIPTION
**Pull request by hacodeorg**
There is a bug in deciding if an application meets minimum requirements using responses from a principal approval form. The issue is reported [here](https://codedotorg.slack.com/archives/G9DF14A56/p1578337769002000). 

The specific question is this one
https://github.com/code-dot-org/code-dot-org/blob/781c8022acc55e6d8ee0ecba1163bd0d5bd5233d/lib/cdo/shared_constants/pd/principal_approval2021_application_constants.rb#L19

which has 4 possible answers.
https://github.com/code-dot-org/code-dot-org/blob/f548aa8d16fe29b835ad83cd922dc6a8d3028fdb/dashboard/app/models/pd/application/principal_approval2021_application.rb#L84-L88

The fix is simple. 
* If the principal answers _Yes_ -> a new course will replaces an existing one -> the application does not meet minimum requirement.
* If the answer is _I don't know_ -> undecided
* All other answers are _No..._ -> the application meets minimum requirement.

## Links
- [Slack discussion](https://codedotorg.slack.com/archives/G9DF14A56/p1578337769002000)
- [Teacher application spec](https://docs.google.com/document/d/1CHjjb7AUUgtX0_deNoxJKeyMjD6DYHjqfQMCjKS1quQ/edit#)

## Testing story
- Adds unit tests to `teacher2021_application_test.rb`.
- Runs `principal_approval2021_application_test.rb`.

## TODO
- [x] Re-score all applications by running `Pd::Application::Teacher2021Application.all.each {|a| a.auto_score!}`

### CODE REVIEW

**Review by agealy — APPROVED** [review]
Thanks for the quick fix!

**Review by islemaster — CHANGES_REQUESTED** [review]
This won't work because the `principal_wont_replace_existing_course` value (which also has a misleading name) is a composed value and won't be an exact match for the `YES` constant.

First, note that this question gets asked on both the teacher application and the principal approval form, that either can impact autoscoring and the principal's answer should "win" if they conflict.

Second, the question text on the principal approval form is:

> Will this course replace an existing computer science course in the master schedule? If yes, please list the course(s) that will be replaced.

Unfortunately the answer we're checking includes that "list the courses" part. Here, after the principal approval is created, the `replace_course`, `replace_course_other`, and `replace_which_course*` fields all get combined and stored as `replace_course_string`

https://github.com/code-dot-org/code-dot-org/blob/f548aa8d16fe29b835ad83cd922dc6a8d3028fdb/dashboard/app/models/pd/application/teacher2021_application.rb#L913-L916

Which is eventually stored as `principal_wont_replace_existing_course`

https://github.com/code-dot-org/code-dot-org/blob/f548aa8d16fe29b835ad83cd922dc6a8d3028fdb/dashboard/app/models/pd/application/teacher2021_application.rb#L948

So I think for this logic to work, you'll want to make sure the principal approval part of the scoring logic checks that the answer starts with yes, not that it's exactly equal.

Please add unit tests for this behavior - the ones that exist don't cover this very well.  Here's an example test case, but it doesn't use a realistic `principal_wont_replace_existing_course` value. I'd like to see a few test cases covering the outcomes you describe in the PR description.

```rb
    test 'autoscore: does not meet minimum requirements if we will replace an existing CS course' do
      application_hash = build :pd_teacher2021_application_hash,
        program: Pd::Application::TeacherApplicationBase::PROGRAMS[:csp],
        replace_existing: YES,
        principal_wont_replace_existing_course: YES

      application = create :pd_teacher2021_application, regional_partner: (create :regional_partner), form_data_hash: application_hash
      application.auto_score!

      assert_equal(
        NO,
        JSON.parse(application.response_scores)['meets_minimum_criteria_scores']['replace_existing']
      )
    end
```

Thanks!

**Comment by hacodeorg** [comment]
Thank you for catching this issue 🤦‍♂ Will fix!

File: dashboard/test/models/pd/application/teacher2021_application_test.rb
Comment (islemaster) [inline dashboard/test/models/pd/application/teacher2021_application_test.rb:794]: What a test!
combo_super_combo_finish

**Review by islemaster — APPROVED** [review]
:+1: Nicely done!

**Pull request merged**
